diff --git a/node_modules/bitcore-lib-cash/index.js b/node_modules/bitcore-lib-cash/index.js
index a562c53..eaa3746 100644
--- a/node_modules/bitcore-lib-cash/index.js
+++ b/node_modules/bitcore-lib-cash/index.js
@@ -12,8 +12,8 @@ bitcore.versionGuard = function(version) {
     throw new Error(message);
   }
 };
-bitcore.versionGuard(global._bitcoreCash);
-global._bitcoreCash = bitcore.version;
+// bitcore.versionGuard(global._bitcoreCash);
+// global._bitcoreCash = bitcore.version;
 
 // crypto
 bitcore.crypto = {};
diff --git a/node_modules/bitcore-lib-cash/lib/crypto/signature.js b/node_modules/bitcore-lib-cash/lib/crypto/signature.js
index 3c9a6f5..1eca0ec 100644
--- a/node_modules/bitcore-lib-cash/lib/crypto/signature.js
+++ b/node_modules/bitcore-lib-cash/lib/crypto/signature.js
@@ -214,14 +214,13 @@ Signature.prototype.toBuffer = Signature.prototype.toDER = function(signingMetho
 
   signingMethod = signingMethod || "ecdsa";
 
+  if(signingMethod === "schnorr") {
+    return Buffer.concat([this.r.toBuffer({size: 32}), this.s.toBuffer({size: 32})]);
+  }
 
   var rnbuf = this.r.toBuffer();
   var snbuf = this.s.toBuffer();
 
-  if(signingMethod === "schnorr") {
-    return Buffer.concat([rnbuf, snbuf]);
-  }
-  
   var rneg = rnbuf[0] & 0x80 ? true : false;
   var sneg = snbuf[0] & 0x80 ? true : false;
 
